Hola **Gloria**!

Soy **Patricio Requena** 👋. Es un placer ser el revisor de tu proyecto el día de hoy!

Revisaré tu proyecto detenidamente con el objetivo de ayudarte a mejorar y perfeccionar tus habilidades. Durante mi revisión, identificaré áreas donde puedas hacer mejoras en tu código, señalando específicamente qué y cómo podrías ajustar para optimizar el rendimiento y la claridad de tu proyecto. Además, es importante para mí destacar los aspectos que has manejado excepcionalmente bien. Reconocer tus fortalezas te ayudará a entender qué técnicas y métodos están funcionando a tu favor y cómo puedes aplicarlos en futuras tareas. 

_**Recuerda que al final de este notebook encontrarás un comentario general de mi parte**_, empecemos!

Encontrarás mis comentarios dentro de cajas verdes, amarillas o rojas, ⚠️ **por favor, no muevas, modifiques o borres mis comentarios** ⚠️:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si todo está perfecto.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>

Puedes responderme de esta forma:
<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class=“tocSkip”></a>
Muchísimas gracias por las observaciones 🤗
</div>

## SQL
El coronavirus tomó al mundo entero por sorpresa, cambiando la rutina diaria de todos y todas. Los habitantes de las ciudades ya no pasaban su tiempo libre fuera, yendo a cafés y centros comerciales; sino que más gente se quedaba en casa, leyendo libros. Eso atrajo la atención de las startups (empresas emergentes) que se apresuraron a desarrollar nuevas aplicaciones para los amantes de los libros.

Te han dado una base de datos de uno de los servicios que compiten en este mercado. Contiene datos sobre libros, editoriales, autores y calificaciones de clientes y reseñas de libros. Esta información se utilizará para generar una propuesta de valor para un nuevo producto.

### Descripción de los datos

**`books`**

Contiene datos sobre libros:

- `book_id`: identificación del libro
- `author_id`: identificación del autor o autora
- `title`: título
- `num_pages`: número de páginas
- `publication_date`: fecha de la publicación
- `publisher_id`: identificación de la editorial

**`authors`**

Contiene datos sobre autores:

- `author_id`: identificación del autor o autora
- `author`: el autor o la autora

**`publishers`**

Contiene datos sobre editoriales:

- `publisher_id`: identificación de la editorial
- `publisher`: la editorial

**`ratings`**

Contiene datos sobre las calificaciones de usuarios:

- `rating_id`: identificación de la calificación
- `book_id`: identificación del libro
- `username`: el nombre del usuario que revisó el libro
- `rating`: calificación

**`reviews`**

Contiene datos sobre las reseñas de los y las clientes:

- `review_id`: identificación de la reseña
- `book_id`: identificación del libro
- `username`: el nombre del usuario que revisó el libro
- `text`: el texto de la reseña

In [1]:
# importar librerías
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'practicum_student',         # nombre de usuario
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # contraseña
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # puerto de conexión
             'db': 'data-analyst-final-project-db'}          # nombre de la base de datos

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Instrucciones del análisis SQL

- Describe los objetivos del estudio.

El objetivo principal de este estudio es analizar los datos registrados sobre los libros que se leyeron durante la pandemia. Esto se logrará a través de consultas SQL para extraer información relevante que sirva de base para diseñar una propuesta de valor atractiva para un nuevo producto digital para las startups.. 


- Estudia las tablas (imprime las primeras filas).

In [2]:
pd.read_sql("SELECT * FROM books LIMIT 5;", con=engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
pd.read_sql("SELECT * FROM authors LIMIT 5;", con=engine)


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [4]:
pd.read_sql("SELECT * FROM publishers LIMIT 5;", con=engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [5]:
pd.read_sql("SELECT * FROM ratings LIMIT 5;", con=engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [6]:
pd.read_sql("SELECT * FROM reviews LIMIT 5;", con=engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

- Realiza una consulta SQL para cada una de las tareas.
- Genera los resultados de cada consulta en el Notebook.
- Describe tus conclusiones para cada una de las tareas.

<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Muy bien, siempre es bueno dar un doble check a los datos con los que se trabajará 
</div>

### Ejercicio
- Encuentra el número de libros publicados después del 1 de enero de 2000.

In [7]:
query = """
SELECT COUNT(*) AS books_2000
FROM books
WHERE publication_date > '2000-01-01';
"""

pd.read_sql(query, con=engine)

books_2000
0         819

Los usuarios registran una preferencia por libros recientes, ya sea por los temas, un lenguaje más actual, accesibilidad o por la popularidad de los libros que se mantienen en la memoria de los usuarios. Podemos hacer una pequeña consulta para comparar los resultados: 

<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Buen trabajo con el cálculo! Aplicaste el filtro correcto para obtener los datos
</div>

In [8]:
query = """
SELECT COUNT(*) AS books_2000
FROM books
WHERE publication_date < '2000-01-01';
"""

pd.read_sql(query, con=engine)

books_2000
0         179

Por lo tanto, los usuarios leen más libros escritos después del 2000. 

- Encuentra el número de reseñas de usuarios y la calificación promedio para cada libro.

In [9]:
query = """
SELECT COUNT(*) AS total_reviews
FROM reviews;
"""

pd.read_sql(query, con=engine)

total_reviews
0           2793

In [10]:
query = """
SELECT 
    books.book_id,
    books.title,
    ROUND(AVG(ratings.rating), 1) AS average_rating
FROM books
JOIN ratings ON books.book_id = ratings.book_id
GROUP BY books.book_id, books.title
ORDER BY book_id ASC
"""
pd.read_sql(query, con=engine)

book_id                                              title  \
0          1                                       'Salem's Lot   
1          2                 1 000 Places to See Before You Die   
2          3  13 Little Blue Envelopes (Little Blue Envelope...   
3          4  1491: New Revelations of the Americas Before C...   
4          5                                               1776   
..       ...                                                ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997                        Xenocide (Ender's Saga  #3)   
997      998                                    Year of Wonders   
998      999                        You Suck (A Love Story  #2)   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...   

     average_rating  
0               3.7  
1               2.5  
2               4.7  
3               4.5  
4               4.0  
..              ...  
995             3.7  
996             3.4  
997             3.2  
998             4.5  
999             3.8  

[1000 rows x 3 columns]

Los usuarios muestran una participación activa en escribir reviews de los libros que leyeron durante la pandemia. Al revisar los libros con mejor calificación, podemos concluir que la audiencia favorece libros de varios tipos de género, tales como misterio por la serie de Women's Club, historia por 1776 y romance por A Bend in the Road. Por lo tanto, los lectores muestran una variedad de historias en sus bibliotecas. 

- Identifica la editorial que ha publicado el mayor número de libros con más de 50 páginas (esto te ayudará a excluir folletos y publicaciones similares de tu análisis).

In [11]:
query = """
SELECT 
    publishers.publisher,
    COUNT(books.book_id) AS cantidad_libros
FROM books
JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE books.num_pages > 50
GROUP BY publishers.publisher
ORDER BY cantidad_libros DESC
LIMIT 1;
"""

pd.read_sql(query, con=engine)

publisher  cantidad_libros
0  Penguin Books               42

La famosa editorial Penguin Books es la que tiene más cantidad de libros. Por lo que se puede recomendar que las start ups procuren algún acuerdo con esta editorial. Igualmente sus libros puede ser los más accesibles para el publico y puede brindar confianza al lector al evaluar la calidad del libro. 

<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Bien utilizadas los diferentes métodos de agregación para obtener los resultados
</div>

- Identifica al autor que tiene la más alta calificación promedio del libro: mira solo los libros con al menos 50 calificaciones.

In [12]:
query = """
SELECT 
  authors.author_id,
  authors.author,
  AVG(ratings.rating) AS best_rating
FROM books 
JOIN authors  ON books.author_id = authors.author_id
JOIN ratings  ON books.book_id = ratings.book_id
JOIN (
    SELECT book_id
    FROM ratings
    GROUP BY book_id
    HAVING COUNT(*) >= 50
) AS popular_books ON books.book_id = popular_books.book_id
GROUP BY authors.author_id, authors.author
ORDER BY best_rating DESC
LIMIT 1;
    
"""

pd.read_sql(query, con=engine)

author_id                      author  best_rating
0        236  J.K. Rowling/Mary GrandPré     4.287097

La autora con libros mejor calificados es J.K. Rowling. Por lo tanto, este podría ser un libro que se mantiene en tendencia por su popularidad o senimiento de nostalgia entre la audiencia. Por lo tanto, sería importante asegurar que la aplicación que se desarrolle considere una lista de series de libros populares. 

<div class="alert alert-block alert-danger">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Aquí deberías hacer uso de subconsultas, una para obtener primero los book_id que tengan mínimo 50 ratings y con ese resultado debes hacer el JOIN con la tabla de authors para sacar el promedio del rating. Sólo deberias mostrar un sólo autor en el resultado.
</div>

- Encuentra el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.

In [13]:
query = """
SELECT 
    AVG(review_counts.review_count) AS avg_num_reviews
FROM (
    SELECT 
        reviews.username,
        COUNT(reviews.review_id) AS review_count
    FROM (
        SELECT 
            username
        FROM ratings
        GROUP BY username
        HAVING COUNT(*) > 50
    ) AS filtered_users
    JOIN reviews ON filtered_users.username = reviews.username
    GROUP BY reviews.username
) AS review_counts;

"""

pd.read_sql(query, con=engine)

avg_num_reviews
0        24.333333

Los usuarios no solo asignan una calificación a los libros, sino que también tienen una preferencia por escribir su opinión sobre el libro. Por lo tanto, este tipo de audiencia procura interactuar, formar una opinión y recomendar a otros los libros que merecen la pena. 

<div class="alert alert-block alert-danger">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Similar al caso anterior, aquí deberías hacer una subconsulta para obtener de la tabla ratings los usuarios con más de 50. Luego con el resultado de esa subconsulta debes filtrar en la tabla reviews para esos usuarios para contar las reviews y con esos resultados en la consulta final sacar el promedio.
    
Debes tener 2 subconsultas y la consulta final que use la información de las otras dos
</div>

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class=“tocSkip”></a>
Gracias Patricio, ya quedó corregido. 
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor (2da Iteracion)</b> <a class=“tocSkip”></a>

Bien! Ahora los resultados cambiaron, hay que tener en cuenta que la forma en cómo traemos los datos va a cambiar el resultado también
</div>

## Conclusión 
Se recomiendan lo siguientes puntos a los start ups para construir sus aplicaciones: 
 - Incluir libros actuales porque son los más leídos por los ususarios. 
 - Planear un acuerdo con las editoriales más populares para asegurar la calidad de los libros y brindar confianza a los lectores. 
 - Integrar una variedad de géneros de libros. Sin embargo, se puede colocar al inicio libros de los géneros más populares como misterio, ficción, romance y aventura. Igulamente al promocionar sus apps, pueden mencionar autores o libros de estos géneros para atraer usuarios. 
 - Procurar también incluir en el catálogo serie de libros populares para asegurar una audiencia inicial. 
 - Los usuarios son una audiencia activa que gusta de participar y recomendar libros. Por lo que es importante incluir una sección de puntaje y una en donde puedan escribir su reseña. 

<div class="alert alert-block alert-success">
<b>Comentario general (2da Iteracion)</b> <a class=“tocSkip”></a>

Buen trabajo con los cambios realizados, ahora realizaste correctamente las consultas para obtener los datos!

Cómo consejo adicional, debido a que SQL es un lenguaje que verás en tu día a día cómo Data Analyst te animo a seguir practicando con el mismo para que en las entrevistas técnicas destaques cómo profesional! Será muy común que cuando postules a un puesto laboral te pidan hacer una prueba de SQL, te recomiendo revisar plataformas cómo Leetcode ya que tienen ejercicios de SQL para prepararte para las entrevistas, particularmente SQL lo encuentras en este link: https://leetcode.com/problemset/database/
    
Saludos!
</div>